src/components/tl/sharp/tl_sharp_context.c

```diff
@@ -126,6 +126,123 @@ static int ucc_tl_sharp_oob_bcast(void *arg, void *buf, int size, int root)
     return status;
 }
 
+static int ucc_tl_sharp_service_barrier(void *arg)
+{
+    ucc_tl_sharp_oob_ctx_t *oob_ctx = (ucc_tl_sharp_oob_ctx_t*)arg;
+    ucc_tl_sharp_context_t *ctx     = (ucc_tl_sharp_context_t*)oob_ctx->ctx;
+    ucc_tl_team_t          *steam   = ctx->super.super.ucc_context->service_team;
+    ucc_subset_t subset;
+    ucc_coll_task_t *req;
+    ucc_status_t status;
+    char sbuf, *rbuf;
+
+
+    subset.map.type   = UCC_EP_MAP_FULL;
+    subset.map.ep_num = oob_ctx->oob->n_oob_eps;
+    subset.myrank     = oob_ctx->oob->oob_ep;
+
+    rbuf = ucc_malloc(sizeof(char) * subset.map.ep_num, "tmp_barrier");
+    if (!rbuf) {
+        tl_error(ctx->super.super.lib,
+                 "failed to allocate %zd bytes for tmp barrier array",
+                 sizeof(char) * subset.map.ep_num);
+        return UCC_ERR_NO_MEMORY;
+    }
+
+    status = UCC_TL_TEAM_IFACE(steam)->scoll.allgather(&steam->super, &sbuf,
+                                                       rbuf, sizeof(char),
+                                                       subset, &req);
+    if (status != UCC_OK) {
+        tl_error(ctx->super.super.lib, "tl sharp gather failed\n");
+        return status;
+    }
+
+    do {
+        ucc_context_progress(ctx->super.super.ucc_context);
+        status = ucc_collective_test(&req->super);
+    } while (status == UCC_INPROGRESS);
+    ucc_collective_finalize(&req->super);
+
+    ucc_free(rbuf);
+
+    return status;
+}
+
+static int ucc_tl_sharp_service_gather(void *arg, int root, void *sbuf,
+                                       void *rbuf, int size)
+{
+    ucc_tl_sharp_oob_ctx_t *oob_ctx  = (ucc_tl_sharp_oob_ctx_t*)arg;
+    ucc_tl_sharp_context_t *ctx      = (ucc_tl_sharp_context_t*)oob_ctx->ctx;
+    ucc_tl_team_t          *steam    = ctx->super.super.ucc_context->service_team;
+    size_t                  msg_size = (size_t)size;
+    ucc_coll_task_t *req;
+    ucc_status_t status;
+    ucc_subset_t subset;
+
+    subset.map.type   = UCC_EP_MAP_FULL;
+    subset.map.ep_num = oob_ctx->oob->n_oob_eps;
+    subset.myrank     = oob_ctx->oob->oob_ep;
+
+    if (subset.myrank != root) {
+        rbuf = ucc_malloc(msg_size * subset.map.ep_num, "tmp_gather");
+        if (!rbuf) {
+            tl_error(ctx->super.super.lib,
+                     "failed to allocate %zd bytes for tmp barrier array",
+                     msg_size * subset.map.ep_num);
+            return UCC_ERR_NO_MEMORY;
+        }
+    }
+
+    status = UCC_TL_TEAM_IFACE(steam)->scoll.allgather(&steam->super, sbuf,
+                                                       rbuf, msg_size,
+                                                       subset, &req);
+    if (status != UCC_OK) {
+        tl_error(ctx->super.super.lib, "tl sharp gather failed\n");
+        return status;
+    }
+
+    do {
+        ucc_context_progress(ctx->super.super.ucc_context);
+        status = ucc_collective_test(&req->super);
+    } while (status == UCC_INPROGRESS);
```

Comment:
i have one concert about this blocking wait here. When we create team (not context) - means e.g. when we create OMPI comm - a process will block there and will not progress runtime (e.g. MPI p2p). I think we can trigger deadlock... @bureddy what do you think? The only right solution i believe is to make sharp_team creation non-blocking similar to UCC. But this will require work.

Reply:
after some debug we couldn't reproduce the hang (looks like additional sync happens in OMPI anyways that prevents that).